### PR TITLE
Use function to remove invalid particles without MPI exchange

### PR DESCRIPTION
**Pull request by RemiLehe**
The new function (introduced by @AlexanderSinn in `amrex`) is a local operation (no MPI communication) and is thus preferable over `ReDistribute` (which also removes invalid particles, but involves MPI communications at the same time).

### CODE REVIEW

**Comment by RemiLehe** [comment]
@roelof-groenewald Do you know whether we have existing CI tests that are sensitive to this part of the code and would confirm that this function is working as expected?

**Comment by roelof-groenewald** [comment]
> Do you know whether we have existing CI tests that are sensitive to this part of the code and would confirm that this function is working as expected?

Probably not since the CI tests were not affected by #4831 and the collision routines have all invalid particles' weights set to 0 (so they don't affect checksum values).
But I can test it locally using the new merging routine since I'm tracking particle numbers in my tests.

File: Source/Particles/WarpXParticleContainer.cpp [hunk 1]
```diff
@@ -302,8 +302,21 @@ WarpXParticleContainer::AddNParticles (int /*lev*/, long n,
 #ifdef AMREX_USE_EB
     auto & distance_to_eb = WarpX::GetInstance().GetDistanceToEB();
     scrapeParticles( *this, amrex::GetVecOfConstPtrs(distance_to_eb), ParticleBoundaryProcess::Absorb());
-    // Call (local) redistribute again to remove particles with invalid ids
-    Redistribute(0, -1, 0, 1, true);
+    deleteInvalidParticles();
```
Comment (roelof-groenewald) [inline Source/Particles/WarpXParticleContainer.cpp:305]: The `#endif` currently on line 320 should be directly below this line. Currently the `WarpXParticleContainer::deleteInvalidParticles` function is only defined for EB=ON compilations.

**Comment by roelof-groenewald** [comment]
@RemiLehe I can confirm that this function works as intended (with the above suggested change included).
I tested a problem where I merge a number of particles on step 2. The case was run with 2 procs. Immediately after merging I print the number of ions using `particle_container.total_number_of_particles(True, True)` (i.e only counting valid particles). I also grab the particle ids and print the length of those arrays as well as the number of invalid entries in the array (i.e. w < 0). The following output is without a `Redistribute` or `deleteInvalidParticles` call:
```
STEP 2 starts ...
--- INFO    : Resampled hydrogen at step 2: macroparticle count decreased by
             516885
Num ions: 123115; len(id) = 320000; invalid ids # 258445
Num ions: 123115; len(id) = 320000; invalid ids # 258440
```
The following output is with a `deleteInvalidParticles` call between merging and the output:
```
STEP 2 starts ...
--- INFO    : Resampled hydrogen at step 2: macroparticle count decreased by
             516885
Num ions: 123115; len(id) = 61555; invalid ids # 0
Num ions: 123115; len(id) = 61560; invalid ids # 0
```

File: Source/Particles/WarpXParticleContainer.cpp [hunk 2]
```diff
@@ -302,8 +302,21 @@ WarpXParticleContainer::AddNParticles (int /*lev*/, long n,
...
+    for (int lev = 0; lev <= nLevels; ++lev) {
+#ifdef AMREX_USE_OMP
+#pragma omp parallel
+#endif
+        for (WarpXParIter pti(*this, lev); pti.isValid(); ++pti) {
+            ParticleTileType& ptile = ParticlesAt(lev, pti);
+            removeInvalidParticles( ptile );
+        }
+    }
 #endif
```
Comment (RemiLehe) [inline Source/Particles/WarpXParticleContainer.cpp]:
```suggestion
```

File: Source/Particles/WarpXParticleContainer.cpp (hunk 1 above)
Comment (RemiLehe) [inline Source/Particles/WarpXParticleContainer.cpp:305]:
```suggestion
    deleteInvalidParticles();
#endif
```

Comment (RemiLehe) [reply Source/Particles/WarpXParticleContainer.cpp:305]: Thanks! Done.

**Review by roelof-groenewald — APPROVED** [review]

**Pull request merged**